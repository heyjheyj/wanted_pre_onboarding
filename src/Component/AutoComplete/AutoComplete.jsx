import React, { useEffect, useState, useCallback, useRef } from "react";
import styles from "./AutoComplete.module.css";

const compareLabel = (a, b) => {
  const labelA = a.label.toUpperCase();
  const labelB = b.label.toUpperCase();

  if (labelA < labelB) return -1;
  if (labelA > labelB) return 1;

  return 0;
};

const AutoComplete = ({ initData }) => {
  const [suggestions, setSuggestions] = useState([]);
  const [text, setText] = useState("");
  const [isAutoCompleting, setIsAutoCompleting] = useState(false);

  const boxRef = useRef();

  const selectSuggestion = useCallback(
    value => {
      setSuggestions([]);
      setText(value);
    },
    [setSuggestions, setText]
  );

  const clear = () => {
    setIsAutoCompleting(false);
    setText("");
    setSuggestions([]);
  };

  const renderSuggestions = useCallback(
    () => {
      return (
        isAutoCompleting &&
        <ul data-testid="test-suggestionlist">
          {suggestions.map((item, index) =>
            <li
              data-testid={`test-suggestion${index}`}
              key={index}
              onClick={() => selectSuggestion(item.label)}
            >
              {item.label}
            </li>
          )}
        </ul>
      );
    },
    [isAutoCompleting, suggestions, selectSuggestion]
  );

  const handleTextChanged = e => {
    let suggestedDataList = [];
    let textValue = e.target.value;

    if (textValue !== "") {
      const tempValue = textValue.replaceAll("\\", "\\\\");
      const regex = new RegExp(`^${tempValue}`, "i");
      suggestedDataList = initData
        .filter(e => regex.test(e.label))
        .sort(compareLabel);
    }
    setSuggestions(suggestedDataList);
    setText(textValue);
  };

  const handleClick = useCallback(
    e => {
      if (boxRef.current && !boxRef.current.contains(e.target)) {
        setIsAutoCompleting(false);
      }
    },
    [setIsAutoCompleting]
  );

  useEffect(
    () => {
      window.addEventListener("mousedown", handleClick);

      return () => {
        window.removeEventListener("mousedown", handleClick);
      };
    },
    [handleClick]
  );

  return (
    <div className={styles.container} data-testid="test-component">
      <h3 className={styles.title}>AutoComplete</h3>
      <div
        className={styles.autoCompleteBox}
        ref={boxRef}
        data-testid="test-container"
      >
        {isAutoCompleting &&
          <button
            data-testid="test-clearbutton"
            className={styles.closebutton}
            onClick={clear}
          >
            x
          </button>}
        <input
          value={text}
          type="text"
          autoComplete="off"
          data-testid="test-input"
          onChange={handleTextChanged}
          onClick={() => setIsAutoCompleting(true)}
        />
        {suggestions.length > 0 && renderSuggestions()}
      </div>
      <span className={styles.text}>Search Top 100 Films</span>
    </div>
  );
};

export default AutoComplete;
