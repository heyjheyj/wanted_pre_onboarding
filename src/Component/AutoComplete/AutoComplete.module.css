.container {
    margin-top: 10px;
    width: 80%;
    height: 200px;
    border: 1px solid grey;
    display: flex;
    flex-direction: column;
    position: relative;
    align-items: center;
    border-radius: 10px;
}

.title {
    position: absolute;
    top: 5px;
    left: 5px;
}

.autoCompleteBox {
    margin-top: 70px;
    width: 80%;
    border: 1px solid grey;
    border-radius: 10px;
    font-size: 14px;
    max-height: 90px;
    overflow: scroll;
    box-shadow: 0 0 0 1px rgba(0,0,0, 0.1), 0 2px 4px 1px rgba(0,0,0, 0.18);
}

.autoCompleteBox > input {
    width: 98%;
    border-radius: 10px;
    padding: 2px 0 2px 5px;
    font-size: 16px;
    border: none;
    box-sizing: border-box;
    outline: none;
}

.autoCompleteBox > ul {
    list-style: none;
    padding: 3px 5px 3px 5px;
    border-top: 1px solid grey;
}

.autoCompleteBox > ul:before {
    content: "";
}

.autoCompleteBox li {
    cursor: pointer;
    padding: 2px 0 2px 0;
}

.autoCompleteBox li:hover {
    background-color: rgb(212, 212, 212);
}

.text {
    position: absolute;
    bottom: 10px;
    color: grey;
}

.closebutton {
    position: absolute;
    right: 12%;
    border: none;
    background-color: transparent;
    margin-top: 2px;
}