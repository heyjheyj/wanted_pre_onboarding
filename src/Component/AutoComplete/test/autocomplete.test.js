import { render, screen } from "@testing-library/react";
import React from "react";
import TestRenderer from "react-test-renderer";
import AutoComplete from "../AutoComplete";
import userEvent from "@testing-library/user-event";

describe("Auto Complete", () => {
  let initData = [
    { label: "Star Wars: Episode V - The Empire Strikes Back", year: 1980 },
    { label: "Forrest Gump", year: 1994 },
    { label: "Inception", year: 2010 },
    { label: "Interstellar", year: 2014 }
  ];
  let component;
  let container;
  let input;
  let suggestionlist;
  let suggestion;
  let clearbutton;

  it("renders", () => {
    const component = TestRenderer.create(<AutoComplete initData={initData} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe("Auto Completing", () => {
    beforeEach(() => {
      render(<AutoComplete initData={initData} />);
      input = screen.getByTestId("test-input");
      container = screen.getByTestId("test-container");
      component = screen.getByTestId("test-component");
    });

    it("shows suggestions when a user types part of search terms on input ", () => {
      userEvent.type(input, "s");

      suggestion = screen.getByTestId("test-suggestion0");
      expect(suggestion.textContent).toBe(
        "Star Wars: Episode V - The Empire Strikes Back"
      );
    });

    it("clears input value and suggestion when 'x'button is clicked", () => {
      userEvent.type(input, "f");

      suggestion = screen.getByTestId("test-suggestion0");
      expect(suggestion.textContent).toBe("Forrest Gump");

      // click button to clear
      clearbutton = screen.getByTestId("test-clearbutton");
      userEvent.click(clearbutton);

      expect(input.value).toBe("");
      expect(
        container.querySelector("button[data-testid=test-clearbutton]")
      ).toBeNull();
      expect(
        container.querySelector("ul[data-testid=test-suggestionlist]")
      ).toBeNull();
    });

    it("hides suggestions when outside of input is clicked", () => {
      userEvent.type(input, "i");
      expect(
        container.querySelector("ul[data-testid=test-suggestionlist]")
      ).toBeDefined();

      // click outside of input and suggestionlist
      container.blur();
      userEvent.click(component);

      expect(input.value).toBe("i");
      expect(
        container.querySelector("ul[data-testid=test-suggestionlist]")
      ).toBeNull();
    });

    it("sets a value in input when a user clicks one of suggestion list", () => {
      userEvent.type(input, "i");
      suggestion = screen.getByTestId("test-suggestion0");
      expect(suggestion.textContent).toBe("Inception");

      // choose a suggestion
      userEvent.click(suggestion);
      expect(input.value).toBe("Inception");
      expect(
        container.querySelector("ul[data-testid=test-suggestionlist]")
      ).toBeNull();
    });

    it("sorts suggestions when suggestions length is more than two", () => {
      userEvent.type(input, "i");
      suggestionlist = screen.getByTestId("test-suggestionlist");

      expect(suggestionlist.childNodes[0].textContent).toBe("Inception");
      expect(suggestionlist.childNodes[1].textContent).toBe("Interstellar");
    });
  });

  describe("dose not sort suggestions when they have same textcontent", () => {
    let props = [
      { label: "Interstellar", year: 2014 },
      { label: "interstellar", year: 2014 }
    ];

    beforeEach(() => {
      render(<AutoComplete initData={props} />);
      input = screen.getByTestId("test-input");
    });

    it("test not to sort", () => {
      userEvent.type(input, "inter");
      suggestionlist = screen.getByTestId("test-suggestionlist");

      expect(suggestionlist.childNodes[0].textContent).toBe("Interstellar");
      expect(suggestionlist.childNodes[1].textContent).toBe("interstellar");
    });
  });
});
