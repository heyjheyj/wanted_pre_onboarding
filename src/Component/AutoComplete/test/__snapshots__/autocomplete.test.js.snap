// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auto Complete renders 1`] = `
<div
  className="container"
  data-testid="test-component"
>
  <h3
    className="title"
  >
    AutoComplete
  </h3>
  <div
    className="autoCompleteBox"
    data-testid="test-container"
  >
    <input
      autoComplete="off"
      data-testid="test-input"
      onChange={[Function]}
      onClick={[Function]}
      type="text"
      value=""
    />
  </div>
  <span
    className="text"
  >
    Search Top 100 Films
  </span>
</div>
`;
